sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoDynamicMessageSchema.java

```diff
@@ -0,0 +1,852 @@
+      FieldDescriptor nanoField = timestampDescriptor.findFieldByNumber(2);
+      long second = (long) timestamp.getField(secondField);
+      int nano = (int) timestamp.getField(nanoField);
+      return Instant.ofEpochSecond(second, nano);
+    }
+
+    @Override
+    void setOnProtoMessage(Message.Builder message, Object value) {
+      if (value != null) {
+        FieldDescriptor fieldDescriptor = getFieldDescriptor(message);
+        message.setField(fieldDescriptor, convertToProtoValue(fieldDescriptor, value));
+      }
+    }
+
+    @Override
+    Object convertToProtoValue(FieldDescriptor fieldDescriptor, Object value) {
+      Row row = (Row) value;
+      return com.google.protobuf.Timestamp.newBuilder()
+          .setSeconds(row.getInt64(0))
+          .setNanos(row.getInt32(1))
+          .build();
+    }
+  }
+
+  static class DurationConvert extends Convert<Object, Object> {
+
+    DurationConvert(Schema.Field field) {
+      super(field);
+    }
+
+    @Override
+    Object getFromProtoMessage(Message message) {
+      FieldDescriptor fieldDescriptor = getFieldDescriptor(message);
+      if (message.hasField(fieldDescriptor)) {
+        Message wrapper = (Message) message.getField(fieldDescriptor);
+        return convertFromProtoValue(wrapper);
+      }
+      return null;
+    }
+
+    @Override
+    Duration convertFromProtoValue(Object object) {
+      Message timestamp = (Message) object;
+      Descriptors.Descriptor timestampDescriptor = timestamp.getDescriptorForType();
+      FieldDescriptor secondField = timestampDescriptor.findFieldByNumber(1);
+      FieldDescriptor nanoField = timestampDescriptor.findFieldByNumber(2);
+      long second = (long) timestamp.getField(secondField);
+      int nano = (int) timestamp.getField(nanoField);
+      return Duration.ofSeconds(second, nano);
+    }
+
+    @Override
+    void setOnProtoMessage(Message.Builder message, Object value) {
+      if (value != null) {
+        FieldDescriptor fieldDescriptor = getFieldDescriptor(message);
+        message.setField(fieldDescriptor, convertToProtoValue(fieldDescriptor, value));
+      }
+    }
+
+    @Override
+    Object convertToProtoValue(FieldDescriptor fieldDescriptor, Object value) {
+      Row row = (Row) value;
+      return com.google.protobuf.Duration.newBuilder()
+          .setSeconds(row.getInt64(0))
+          .setNanos(row.getInt32(1))
+          .build();
+    }
+  }
+
+  static class MessageConvert extends Convert<Object, Object> {
+    private final SerializableFunction fromRowFunction;
+    private final SerializableFunction toRowFunction;
+
+    MessageConvert(ProtoDynamicMessageSchema rootProtoSchema, Schema.Field field) {
+      super(field);
+      ProtoDynamicMessageSchema protoSchema =
+          ProtoDynamicMessageSchema.forContext(rootProtoSchema.context, field);
+      SchemaCoder<DynamicMessage> schemaCoder = protoSchema.getSchemaCoder();
```

Comment:
I don't think we should be storing SchemaCoder objects around here. I think preferable to store schema and to/from row functions separately.

Reply:
I've changed the accesses here, but iI still think that it's ok to use SchemaCoder as a transient container for the trio: Schema, ToRow and FromRow.

If it's a blocker, I can spit it up in 3 different transient fields.

Reply:
I would prefer 3 fields. SchemaCoder is not a container object - it just happens to contain those fields. In fact I would prefer to make SchemaCoder private (so that users of Schemas don't have to ever deal with coders at all), however that would require more major changes to Beam.

Reply:
Removed coder as container.